src/Setup_geometry.jl

```diff
@@ -1177,3 +1179,354 @@ function Compute_ThermalStructure(Temp, X, Y, Z, Phase, s::LinearWeightedTempera
+
+            d[ind_seg[ip]] = ScatteredInterpolation.evaluate(itp1,point_)[1];
+            ls[ind_seg[ip]] = ScatteredInterpolation.evaluate(itp2,point_)[1];
+        end
+
+        #Update l
+        l = ln;
+    end
+end
+
+"""
+    addSlab!(X::Array{Float64},Y::Array{Float64},Z::Array{Float64},Ph::Array{Int32},T::Array{Float64},t::Trench,strat,temp)
+
+Main function that creates the slab. Unpack the variable from the structure, create two arrays that contains the information of l and d.
+    l and d are the array containing the length of the slab per each coordinate belonging to the slab, and the distance from the surface.
+    In this function compute_slab_surface and find_slab are called. And after d and ls are computed, it fill up the temperature and phase arrays
+
+"""
+function addSlab!(X::Array{Float64},Y::Array{Float64},Z::Array{Float64},Ph::Array{Int32},T::Array{Float64},t::Trench,strat,temp)
+
+    d = ones(size(X)).*NaN64;
+
+    # -> l = length from the trench along the slab
+    ls = ones(size(X)).*NaN64;
+
+    D0 = t.D0;
+
+    L0 = t.L0;
+
+    n_seg = t.n_seg;
+
+    Lb    = t.Lb;
+
+    θ_max = t.θ_max;
+
+    n_seg_xy = t.n_seg_xy;
+
+    #WZ = t.WZ;
+
+    # Allocate d-l array and A,B
+
+    A = zeros(Float64,2,1);
+    B = zeros(Float64,2,1);
+
+    #Loop over the segment of the slab
+    # In theory this loop would loop all the segment of the trench, but if I introduce a n_seg_xy = 1, it throws me an error concerning the iteration. I tried to fix, but the the error message is mysterious
+
+    #for is =1:1
+        is = 1;
+        A[1] = t.A[is];
+        A[2] = t.A[is+1];
+        B[1] = t.B[is];
+        B[2] = t.B[is+1];
+
+        # Compute Top-Bottom surface
+        # Or loop over the segment of the top/bottom surface and inpolygon each element or
+
+        Top,Bottom =compute_slab_surface!(D0,L0,Lb,n_seg,abs(θ_max),t.type_bending);
+
+        find_slab!(X,Y,Z,d,ls,t.θ_max,A,B,Top,Bottom,t.n_seg,t.D0,t.L0,t.direction);
+
+        l_decouplingind = findall(Top[:,2].<=-t.d_decoupling);
+
+        l_decoupling = Top[l_decouplingind[1],1];
+
+        # Function to fill up the temperature and the phase. I need to personalize addbox!
+
+        ind = findall((-D0 .<= d .<= 0.0));
+
+        if typeof(temp) == LinearWeightedTemperature
+            l_decouplingind = findall(Top[:,2].<=-t.d_decoupling);
+
+            l_decoupling = Top[l_decouplingind[1],1];
+    
+            temp.crit_dist = l_decoupling; 
+
+        end
+        # Compute thermal structure accordingly. See routines below for different options {Future: introducing the length along the trench for having lateral varying properties along the trench}
+        T[ind] = Compute_ThermalStructure(T[ind], ls[ind], Y[ind], d[ind],Ph[ind],temp);
+
+        # Set the phase. Different routines are available for that - see below.
+        Ph[ind] = Compute_Phase(Ph[ind], T[ind], ls[ind], Y[ind], d[ind], strat)
```

Comment:
Every `T[ind]` (or similar) that you do, allocates an array. If you move this to a loop, you would be doing `T[1]` for example, that does not allocate anything.

Reply:
I have not tried to run your code, but refactoring in a way that you don't allocate these intermediate arrays would improve performance and obviously get rid of memory allocations

Reply:
However, in the context of this function I need to have an index array to pass selectively part of Temp and Phase to compute temperature/phase field. I do not know if I can use these functions only for points. 
Ideally having a loop would allow more flexibility, for example, introducing variation of the angle of subduction along strike, temperature field, thickness and length. 
For now I can keep this formulation, and after the pull request discuss with Boris about that for upgrading these functions for further developments.

Reply:
I see. Yes ideally you would have versions of these functions that operate locally, that would make it more flexible and able to fuse operations into single loops when possible